katdal/chunkstore_s3.py

```diff
@@ -192,35 +237,61 @@ def get_chunk(self, array_name, slices, dtype):
     def put_chunk(self, array_name, slices, chunk):
         """See the docstring of :meth:`ChunkStore.put_chunk`."""
         chunk_name, _ = self.chunk_metadata(array_name, slices, chunk=chunk)
-        bucket, key = self.split(chunk_name + '.npy', 1)
+        url = self._chunk_url(chunk_name)
         fp = io.BytesIO()
         np.lib.format.write_array(fp, chunk, allow_pickle=False)
+        md5 = base64.b64encode(hashlib.md5(fp.getvalue()).digest())
         fp.seek(0)
-        with self._standard_errors(chunk_name):
-            self.client.put_object(Bucket=bucket, Key=key, Body=fp)
+        headers = {'Content-MD5': md5}
+        with self._standard_errors(chunk_name), self._session_pool() as session:
+            with contextlib.closing(session.put(url, headers=headers, data=fp)) as response:
+                _raise_for_status(response)
 
     def has_chunk(self, array_name, slices, dtype):
         """See the docstring of :meth:`ChunkStore.has_chunk`."""
         dtype = np.dtype(dtype)
         chunk_name, _ = self.chunk_metadata(array_name, slices, dtype=dtype)
-        bucket, key = self.split(chunk_name + '.npy', 1)
-        with self._standard_errors(chunk_name):
-            try:
-                self.client.head_object(Bucket=bucket, Key=key)
-            except ClientError as err:
-                if err.response['Error']['Code'] != '404':
-                    raise
-                return False
-            else:
-                return True
+        url = self._chunk_url(chunk_name)
+        try:
+            with self._standard_errors(chunk_name), self._session_pool() as session:
+                with contextlib.closing(session.head(url)) as response:
+                    _raise_for_status(response)
+        except ChunkNotFound:
+            return False
+        else:
+            return True
+
+    list_max_keys = 100000
 
     def list_chunk_ids(self, array_name):
         """See the docstring of :meth:`ChunkStore.list_chunk_ids`."""
+        NS = '{http://s3.amazonaws.com/doc/2006-03-01/}'
         bucket, prefix = self.split(array_name, 1)
-        paginator = self.client.get_paginator('list_objects')
-        page_iter = paginator.paginate(Bucket=bucket, Prefix=prefix,
-                                       PaginationConfig={'PageSize': 10000})
-        keys = [item['Key'] for page in page_iter for item in page['Contents']]
+        url = urlparse.urljoin(self._url, urllib.quote(bucket))
+        params = {
+            'prefix': prefix,
+            'max-keys': self.list_max_keys
+        }
+
+        keys = []
+        more = True
+        while more:
+            with self._standard_errors(), self._session_pool() as session:
+                with contextlib.closing(session.get(url, params=params)) as response:
+                    _raise_for_status(response)
```

Comment:
I'm wondering whether the `with... with... _raise_for_status...` triplet of lines can be rolled into a single context manager since it sees a lot of use, but then every time it is subtly different due to the specific session method etc. Mmmm, thoughts?

Reply:
Nice idea. I've implemented it as `self._request`.